src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/Marshalling/Utf8StringMarshaller.cs

```diff
@@ -10,93 +10,112 @@ namespace System.Runtime.InteropServices.Marshalling
         /// </summary>
-        /// <param name="str">The string to marshal.</param>
-        /// <param name="buffer">Buffer that may be used for marshalling.</param>
-        /// <remarks>
-        /// The <paramref name="buffer"/> must not be movable - that is, it should not be
-        /// on the managed heap or it should be pinned.
-        /// </remarks>
-        /// <seealso cref="CustomTypeMarshallerFeatures.CallerAllocatedBuffer"/>
-        public Utf8StringMarshaller(string? str, Span<byte> buffer)
+        public ref struct ManagedToUnmanagedIn
         {
-            _allocated = false;
+            /// <summary>
+            /// Requested buffer size for optimized marshalling.
+            /// </summary>
+            public static int BufferSize { get; } = 0x100;
+
+            private byte* _unmanagedValue;
+            private bool _allocated;
 
-            if (str is null)
+            /// <summary>
+            /// Initialize the marshaller with a managed string and requested buffer.
+            /// </summary>
+            /// <param name="managed">The managed string</param>
+            /// <param name="buffer">A request buffer of at least size, <see cref="BufferSize"/>.</param>
+            public void FromManaged(string? managed, Span<byte> buffer)
             {
-                _nativeValue = null;
-                return;
-            }
+                _allocated = false;
 
-            const int MaxUtf8BytesPerChar = 3;
+                if (managed is null)
+                {
+                    _unmanagedValue = null;
+                    return;
+                }
 
-            // >= for null terminator
-            // Use the cast to long to avoid the checked operation
-            if ((long)MaxUtf8BytesPerChar * str.Length >= buffer.Length)
-            {
-                // Calculate accurate byte count when the provided stack-allocated buffer is not sufficient
-                int exactByteCount = checked(Encoding.UTF8.GetByteCount(str) + 1); // + 1 for null terminator
-                if (exactByteCount > buffer.Length)
+                const int MaxUtf8BytesPerChar = 3;
+
+                // >= for null terminator
+                // Use the cast to long to avoid the checked operation
+                if ((long)MaxUtf8BytesPerChar * managed.Length >= buffer.Length)
                 {
-                    buffer = new Span<byte>((byte*)Marshal.AllocCoTaskMem(exactByteCount), exactByteCount);
-                    _allocated = true;
+                    // Calculate accurate byte count when the provided stack-allocated buffer is not sufficient
+                    int exactByteCount = checked(Encoding.UTF8.GetByteCount(managed) + 1); // + 1 for null terminator
+                    if (exactByteCount > buffer.Length)
+                    {
+                        buffer = new Span<byte>((byte*)Marshal.AllocCoTaskMem(exactByteCount), exactByteCount);
```

Comment:
This can use `NativeMemory.Alloc` now that the `In` mode is separate from the rest.  `NativeMemory.Alloc` is faster than `Marshal.AllocCoTaskMem`.

Reply:
(same for other marshallers)